### DISCUSSION TITLE
How to start the HTML response stream while `getInitialProps` is in progress?

### DISCUSSION THREAD

**Discussion by mohsen1** [vercel/next.js#16730] [Ideas]
Our server is kind of slow. Doing SSR means calling the slow server and waiting for the data before being able to send bytes down the pipe. 

<img width="488" alt="Screen Shot 2020-08-31 at 3 21 53 PM" src="https://user-images.githubusercontent.com/543633/91758304-c6726500-eb9d-11ea-9cad-1a3511649e56.png">

Is there a mechanism to start streaming HTML response before `getInitialProps` is completed? Maybe HTTP/2 Push can help?

**Comment by mohsen1** [comment]
I made some changes to the HTTP2 example to further illustrate this problem:

https://github.com/vercel/next.js/compare/canary...mohsen1:http2-push?expand=1


Screenshot from the example above:

<img width="841" alt="Screen Shot 2020-08-31 at 5 54 59 PM" src="https://user-images.githubusercontent.com/543633/91772577-1e679680-ebb3-11ea-9f1f-4747130cd332.png">

**Comment by stfnsr** [comment]
Have you considered making the page static, and fetching the required data afterwards from an api route?

**Reply by mohsen1** [reply]
That has a worse performance characteristics because it relies on client to fully bootstrap and then initiates the XHR call.

**Comment by mohsen1** [comment]
I think HTTP/2 Push is an overkill for this. All we need to do is to start streaming the HTML HTTP response by sending the document `<head>` before getInitialProps is completed.

**Comment by mohsen1** [comment]
I wrote an RFC to implement eager `<head>` to solve this problem.

https://github.com/vercel/next.js/discussions/16854
